src/modules/silentpayments/main_impl.h

```diff
@@ -0,0 +1,629 @@
+    secp256k1_pubkey pubkey;
+    unsigned char input_hash[32];
+    size_t pubkeylen = 33;
+    int ret = 1;
+
+    VERIFY_CHECK(ctx != NULL);
+    ARG_CHECK(output33 != NULL);
+    ARG_CHECK(public_data != NULL);
+    /* Only allow public_data to be serialized if it has the hash and the summed public key
+     * This helps protect against accidentally serialiazing just a the summed public key A
+     */
+    ARG_CHECK(public_data->data[0] == 0);
+    ret &= secp256k1_silentpayments_recipient_public_data_load_pubkey(ctx, &pubkey, public_data);
+    secp256k1_silentpayments_recipient_public_data_load_input_hash(input_hash, public_data);
+    ret &= secp256k1_ec_pubkey_tweak_mul(ctx, &pubkey, input_hash);
+    secp256k1_ec_pubkey_serialize(ctx, output33, &pubkeylen, &pubkey, SECP256K1_EC_COMPRESSED);
+    return ret;
+}
+
+int secp256k1_silentpayments_recipient_public_data_parse(const secp256k1_context *ctx, secp256k1_silentpayments_public_data *public_data, const unsigned char *input33) {
+    size_t inputlen = 33;
+    size_t pubkeylen = 65;
+    secp256k1_pubkey pubkey;
+
+    VERIFY_CHECK(ctx != NULL);
+    ARG_CHECK(public_data != NULL);
+    ARG_CHECK(input33 != NULL);
+    /* Since an attacker can send us malicious data that looks like a serialized public key but is not, fail early */
+    if (!secp256k1_ec_pubkey_parse(ctx, &pubkey, input33, inputlen)) {
+        return 0;
+    }
+    public_data->data[0] = 1;
+    secp256k1_ec_pubkey_serialize(ctx, &public_data->data[1], &pubkeylen, &pubkey, SECP256K1_EC_UNCOMPRESSED);
+    memset(&public_data->data[1 + pubkeylen], 0, 32);
+    return 1;
+}
+
+int secp256k1_silentpayments_recipient_scan_outputs(
+    const secp256k1_context *ctx,
+    secp256k1_silentpayments_found_output **found_outputs, size_t *n_found_outputs,
+    const secp256k1_xonly_pubkey * const *tx_outputs, size_t n_tx_outputs,
+    const unsigned char *recipient_scan_key,
+    const secp256k1_silentpayments_public_data *public_data,
+    const secp256k1_pubkey *recipient_spend_pubkey,
+    const secp256k1_silentpayments_label_lookup label_lookup,
+    const void *label_context
+) {
+    secp256k1_scalar t_k_scalar, rsk_scalar;
+    secp256k1_ge label_ge, recipient_spend_pubkey_ge, A_sum_ge;
+    secp256k1_pubkey A_sum;
+    secp256k1_xonly_pubkey P_output_xonly;
+    unsigned char shared_secret[33];
+    const unsigned char *label_tweak = NULL;
+    size_t i, k, n_found, found_idx;
+    int found, combined;
+    int ret = 1;
+
+    /* Sanity check inputs */
+    VERIFY_CHECK(ctx != NULL);
+    ARG_CHECK(found_outputs != NULL);
+    ARG_CHECK(n_found_outputs != NULL);
+    ARG_CHECK(tx_outputs != NULL);
+    ARG_CHECK(n_tx_outputs > 0);
+    ARG_CHECK(recipient_scan_key != NULL);
+    ARG_CHECK(public_data != NULL);
+    ARG_CHECK(recipient_spend_pubkey != NULL);
+    if (label_lookup != NULL) {
+        ARG_CHECK(label_context != NULL);
+    } else {
+        ARG_CHECK(label_context == NULL);
+    }
```

Comment:
Why is the label_context ARG_CHECK'ed, shouldn't this be treated as opaque data for the callback alone?

Reply:
That seems unnecessary, as both the callback and the context pointer are passed by the caller. If the caller wants to use global state or some other way that does not require `label_context` it should be allowed to set it to `NULL`.

Reply:
Since the label_lookup accepts a void* pointer for the context, I think it would be fine to make passing the context optional. Perhaps the better arg check would be to do this in reverse: complain if the caller passes a context and not a lookup function. Will update.